### DISCUSSION TITLE
about edge_index in Data

### DISCUSSION THREAD

**Discussion by ZIKEYUAN** [pyg-team/pytorch_geometric#2300] [Q&A]
In the Data file, edge_index always contains twice as much data as the existing edges. Can I understand that the Data dataset splits the edges in the undirected graph into two back and forth edges? In this way, can we continue to understand that edge_index already contains the direction information of the edge?
The second question is, if the graph I'm dealing with is a directed graph, I put the edge information into edge_index, so that I don't need to double the edge data?
Because the research I am doing now is about directed graphs, I would like to use PYG to construct the GNN network structure I designed, but I encountered a big problem with edge_index because I added edge_type to the data of undirected graphs. After the data (that is, the type of each edge is marked as 0), the final training result obtained is very different from that without edge_type. In theory, there should be no difference. I really want to solve this problem.

**Comment by rusty1s** [comment]
Copying from my mail:

(1) you are absolutely right. `edge_index` decodes the direction of an edge, and in order to represent undirected graphs, we need to add both directions to `edge_index`.

(2) I'm not entirely sure to understand your questions, but even when operating in directed graphs, it's always a good idea to let the network know about reverse edges. Otherwise, the message passing flow is only unidirectional. In order to distinguish those edges, one could make use of separate transformation matrices for "real" and "reverse" edges.
